include/envoy/thread_local/thread_local.h

```diff
@@ -46,6 +46,14 @@ class Slot {
    */
   virtual void runOnAllThreads(Event::PostCb cb) PURE;
 
+  /**
+   * Run a callback on all registered threads with a barrier. There is no guarantee that the
+   * callback be fired during shutdown.
+   * @param cb supplies the callback to run on each thread.
+   * @param main_callback supplies the callback to run on main thread after threads are done.
```

Comment:
s/main_callback/all_threads_complete_cb (or something). Also per above clarify this runs on the main thread. You should also ASSERT in the impl that this function is only called on the main thread.